src/framework/serialize.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * serialize server rendering result
+ * 1. replace symbols
+ * 2. replace client component with placeholder
+ */
+export default function serialize(json) {
+  const replaced = replaceClientComponent(json);
+
+  return JSON.stringify(replaced, (k, v) => {
+    if (k === "$$typeof" && typeof v === "symbol") {
+      return v.toString();
+    }
+    return v;
+  });
+}
+
+function replaceClientComponent(data) {
+  if (data == null || typeof data !== "object") {
+    return data;
+  }
+
+  if (Array.isArray(data)) {
+    return data.map(replaceClientComponent);
+  }
+
+  // if it is client component
+  // switch it to LazyContainer
+  // we assume all function components are client components, which clearly is not true
+  // TODO: fix this
+  if (
+    data.$$typeof === Symbol.for("react.element") &&
+    typeof data.type === "function"
+  ) {
+    return {
+      ...data,
+      props: {
+        ...replaceClientComponent(data.props),
+        // TODO: key conflict
+        componentName: data.type.name,
+      },
+      type: "$LazyContainer",
```

Comment:
we changed the `type` to a placeholder string, since function is not serializable. 
it is revived during deserializing.